dashboard/app/jobs/application_job.rb

```diff
@@ -1,7 +1,72 @@
+require 'cdo/aws/metrics'
+
 class ApplicationJob < ActiveJob::Base
   # Automatically retry jobs that encountered a deadlock
   # retry_on ActiveRecord::Deadlocked
 
   # Most jobs are safe to ignore if the underlying records are no longer available
   # discard_on ActiveJob::DeserializationError
+
+  METRICS_NAMESPACE = 'code-dot-org/ActiveJob'
+
+  def self.common_dimensions
+    [
+      {
+        name: 'Environment',
+        value: CDO.rack_env
+      },
+      {
+        name: 'JobName',
+        value: name
+      }
+    ]
+  end
+
+  # Parent class callbacks are called in addition to any callbacks defined in the job subclass
+  # https://guides.rubyonrails.org/v6.0/active_job_basics.html#available-callbacks
+
+  before_perform do |job|
+    @perform_started_at = Time.now
+    wait_time = @perform_started_at - Time.parse(job.enqueued_at)
+
+    metric = {
+      metric_name: 'WaitTime',
+      value: wait_time,
+      unit: 'Seconds',
+      dimensions: job.class.common_dimensions,
+    }
+
+    pp metric
+    Cdo::Metrics.put_metric(METRICS_NAMESPACE, metric)
+  end
+
+  after_perform do |job|
+    perform_complete_at = Time.now
+    execution_time = perform_complete_at - @perform_started_at
+    total_time = perform_complete_at - Time.parse(job.enqueued_at)
+    jobs_in_queue = Delayed::Job.where(queue: job.queue_name).where("handler LIKE ?", "%job_class: #{job.class.name}%").count
```

Comment:
This is pretty ugly, and perhaps not very performant. Very interested in other ideas here. I would love to keep this as as per-job-class metric, but we could simplify by just reporting the total queue size with `Delayed::Job.count`

Reply:
This comment is marked as outdated because I moved this line.

Reply:
if I understand correctly, you are trying to monitor the activejob infrastructure, and not the implementation details of any particular job. based on that, monitoring/alerting on total # jobs seems best to me.

in the future if we end up wanting to monitor queues for individual jobs (for example if we define more than one queue), then perhaps we would shard the monitoring based on the `queue` field rather than inspecting the handler.

Reply:
Sounds good to me. We can always query the database if we have a non-zero number we want to learn more about.